src/cql-code-service.ts

```diff
@@ -1,32 +1,34 @@
-const { Code, ValueSet } = require('./datatypes/datatypes');
+import { Code, ValueSet } from './datatypes/datatypes';
 
-class CodeService {
-  constructor(valueSetsJson = {}) {
+export class CodeService {
+  valueSets: any;
+
+  constructor(valueSetsJson: any = {}) {
     this.valueSets = {};
-    for (let oid in valueSetsJson) {
+    for (const oid in valueSetsJson) {
       this.valueSets[oid] = {};
-      for (let version in valueSetsJson[oid]) {
+      for (const version in valueSetsJson[oid]) {
         const codes = valueSetsJson[oid][version].map(
-          code => new Code(code.code, code.system, code.version)
+          (code: any) => new Code(code.code, code.system, code.version)
         );
         this.valueSets[oid][version] = new ValueSet(oid, version, codes);
       }
     }
   }
 
-  findValueSetsByOid(oid) {
+  findValueSetsByOid(oid: string) {
```

Comment:
Hopefully (at least partially) resolved in [b215c5a](https://github.com/cqframework/cql-execution/pull/260/commits/b215c5ab0b1faa433a5982562fe7ff5b9164a707)

Let me know if this is what you had in mind.

Reply:
Going for the bonus point!  I like it!

I'll review the code more closely in a 2nd pass, but I think this is the basic direction I was looking to go.  The one thing I noticed is that you introduced the interfaces (for `TerminologyProvider` and `DataProvider`, and made the `class` implementations `implement` them, but you did not change the functions in `Executor` that take them as arguments.

If you can update the functions/constructor in `Executor` that takes `CodeService` and `PatientSource` as arguments and swap out the argument types to use the interfaces, then that would be more clear to implementors -- and it would prove to you if you actually got the interfaces right (i.e., you'll get compile errors if you defined the interfaces wrong).

Reply:
Ah yes, good call. Hopefully addressed in [5785caa](https://github.com/cqframework/cql-execution/pull/260/commits/5785caafead6072bb73c57364017b556c7d213f6)